File: testdata/fixtures/typescript/CWE-89.ts
// CWE-89: SQLi via template
import { Request } from 'express';

function vuln(req: Request, conn: any) {
  const id = req.query.get("id") as string;
  // ruleid: typescript-sqli
  conn.query(`SELECT * FROM users WHERE id = ${id}`);
  const name = req.params.get("name") as string;
  // ruleid: typescript-sqli
  conn.execute(`SELECT * FROM t WHERE name = '${name}'`);
  const q = location.search;
  // ruleid: typescript-sqli
  db.query(q);
}
function safe(req: Request, conn: any) {
  // ok: typescript-sqli
  conn.query("SELECT * FROM users WHERE id = ?", [1]);
  // ok: typescript-sqli
  conn.execute("SELECT 1");
}
declare const db: any;

File: testdata/fixtures/scala/CWE-89.scala
// CWE-89: SQL injection via JDBC — string from request reaches SQL execution
import java.sql.{Connection, DriverManager}
import javax.servlet.http.HttpServletRequest

object SqlTest {
  def vuln(request: HttpServletRequest, conn: Connection) = {
    val id = request.getParameter("id")
    val name = request.getHeader("X-Name")
    // ruleid: scala-sql-injection
    conn.prepareStatement("SELECT * FROM users WHERE id = " + id).executeQuery()
    // ruleid: scala-sql-injection
    conn.createStatement().executeQuery("SELECT * FROM t WHERE name = '" + name + "'")
    val q = request.getQueryString()
    // ruleid: scala-sql-injection
    conn.prepareStatement(q).execute()
  }
  def safe(request: HttpServletRequest, conn: Connection) = {
    // ok: scala-sql-injection
    conn.prepareStatement("SELECT * FROM users WHERE id = ?").executeQuery()
    // ok: scala-sql-injection
    val fixed = "SELECT * FROM t"
    conn.prepareStatement(fixed).executeQuery()
    // ok: scala-sql-injection
    conn.prepareStatement("SELECT 1").execute()
  }
}
